federation/pkg/federation-controller/namespace/namespace_controller.go

```diff
@@ -0,0 +1,305 @@
+	nc.namespaceDeliverer = util.NewDelayingDeliverer()
+	nc.clusterDeliverer = util.NewDelayingDeliverer()
+
+	// Start informer in federated API servers on namespaces that should be federated.
+	nc.namespaceInformerStore, nc.namespaceInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Core().Namespaces().List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Core().Namespaces().Watch(options)
+			},
+		},
+		&api_v1.Namespace{},
+		controller.NoResyncPeriodFunc(),
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { nc.deliverNamespaceObj(obj, 0, false) }))
+
+	// Federated informer on namespaces in members of federation.
+	nc.namespaceFederatedInformer = util.NewFederatedInformer(
+		client,
+		func(cluster *federation_api.Cluster, targetClient federation_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Core().Namespaces().List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Core().Namespaces().Watch(options)
+					},
+				},
+				&api_v1.Namespace{},
+				controller.NoResyncPeriodFunc(),
+				// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+				// would be just confirmation that some namespace opration suceeded.
+				util.NewTriggerOnMetaAndSpecChangesPreproc(
+					func(obj pkg_runtime.Object) { nc.deliverNamespaceObj(obj, nc.namespaceReviewDelay, false) },
+					func(obj pkg_runtime.Object) { util.SetClusterName(obj, cluster.Name) },
+				))
+		},
+
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the namespaces again.
+				nc.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(nc.clusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	nc.federatedUpdater = util.NewFederatedUpdater(nc.namespaceFederatedInformer,
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Create(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			_, err := client.Core().Namespaces().Update(namespace)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			namespace := obj.(*api_v1.Namespace)
+			err := client.Core().Namespaces().Delete(namespace.Name, &api.DeleteOptions{})
+			return err
+		})
+	return nc
+}
+
+func (nc *NamespaceController) Start() {
```

Comment:
For consistency with all other existing controllers, can you call it "Run" and take "stopCh <-chan struct{}) as parameter?